lib/model.py

```diff
@@ -1,45 +1,73 @@
-from typing import List
+from typing import List, Tuple, Union
 
+import numpy as np
+import pandas as pd
 import tensorflow as tf
 
 
-@tf.keras.utils.register_keras_serializable()
-class OutputMapperLayer(tf.keras.layers.Layer):
-    """
-    The OutputMapperLayer converts the label indices produced by the model to
-    the taxonomy category ids and limits them to top N labels.
+@tf.function
+def top_labeled_predictions(
+        predictions: Union[tf.Tensor, np.array],
+        labels: List[str],
+        k: int = 10):
     """
+    Top labeled predictions.
+
+    This `@tf.function` can be used as a custom serving function.
+
+    Parameters
+    ----------
+    predictions: tf.Tensor or np.array
+        Predictions, as returned by `model.predict` or equivalent.
+
+    labels: List[str]
+        Label vocabulary.
 
-    def __init__(self, labels: List[str], top_n: int, **kwargs):
-        self.labels = labels
-        self.top_n = top_n
+    k: int, optional
+        Number of top predictions to return.
 
-        super(OutputMapperLayer, self).__init__(**kwargs)
+    Returns
+    -------
+    {'labels': tf.Tensor, 'scores': tf.Tensor}
+        Top predicted labels with their scores.
+        Returned tensors will have shape `(predictions.shape[0], k)`.
+    """
+    batch_size = tf.shape(predictions)[0]
+    tf_labels = tf.constant([labels], dtype='string')
+
+    top_indices = tf.nn.top_k(predictions, k=k, sorted=True, name='top_k').indices
 
-    def call(self, x):
-        batch_size = tf.shape(x)[0]
+    top_labels = tf.experimental.numpy.take(tf_labels, top_indices)
+    top_scores = tf.gather(predictions, top_indices, batch_dims=1)
 
-        tf_labels = tf.constant([self.labels], dtype="string")
-        tf_labels = tf.tile(tf_labels, [batch_size, 1])
+    return {'labels': top_labels, 'scores': top_scores}
 
-        top_n = tf.nn.top_k(x, k=self.top_n, sorted=True, name="top_k").indices
 
-        top_conf = tf.gather(x, top_n, batch_dims=1)
-        top_labels = tf.gather(tf_labels, top_n, batch_dims=1)
+def top_predictions_table(labeled_predictions) -> pd.DataFrame:
+    """
+    Format the top labeled predictions into a pretty table.
 
-        return (top_conf, top_labels)
+    Parameters
+    ----------
+    labeled_predictions: {'labels': tf.Tensor, 'scores': tf.Tensor}
+        Labeled predictions, as returned by `top_labeled_predictions`.
 
-    def compute_output_shape(self, input_shape):
-        batch_size = input_shape[0]
-        top_shape = (batch_size, self.top_n)
-        return [top_shape, top_shape]
+    Returns
+    -------
+    pd.DataFrame
+    """
+    labels = pd.DataFrame(labeled_predictions['labels'].numpy()).stack()
+    scores = pd.DataFrame(labeled_predictions['scores'].numpy()).stack()
```

Comment:
i'm absolutely fine with this, but I find it "cleaner"  to do all the job with numpy before converting to a Pd.DataFrame rather than going back and forth between pandas and numpy.
```suggestion
labels = labeled_predictions['labels'].numpy()
scores = labeled_predictions['scores'].numpy()
```

Reply:
Thanks for the suggestion! I also find it cleaner :)
I took your proposed changes (below as well) and refactored the two `np.vectorize` into a single call. I find it even more explicit and it's also slightly faster.